workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourLibrary.cs

```diff
@@ -0,0 +1,166 @@
+        }
+
+        public IEnumerable<uint> GetRequiredWriterComponentIds(Type behaviourType)
+        {
+            EnsureLoaded(behaviourType);
+            return componentWriterIdsForBehaviours[behaviourType];
+        }
+
+        private void Inject(MonoBehaviour spatialOSBehaviour, uint componentId)
+        {
+            var readerWriter = ReaderWriterFactory.CreateReaderWriter(componentId);
+            var memberAdapter = adapterCache[spatialOSBehaviour.GetType()][componentId];
+            memberAdapter.SetValue(spatialOSBehaviour, readerWriter);
+        }
+
+        private void DeInject(MonoBehaviour spatialOSBehaviour, uint componentId)
+        {
+            var memberAdapter = adapterCache[spatialOSBehaviour.GetType()][componentId];
+            memberAdapter.SetValue(spatialOSBehaviour, null);
+        }
+
+        private void EnsureLoaded(Type behaviourType)
+        {
+            if (adapterCache.ContainsKey(behaviourType))
+            {
+                return;
+            }
+
+            var adapters = reflectionUtil.GetMembersWithMatchingAttributes(behaviourType);
+            var componentIdsToAdapters = new Dictionary<uint, IMemberAdapter>();
+            var readerIds = new List<uint>();
+            var writerIds = new List<uint>();
+            adapterCache[behaviourType] = componentIdsToAdapters;
+            componentReaderIdsForBehaviours[behaviourType] = readerIds;
+            componentWriterIdsForBehaviours[behaviourType] = writerIds;
+            foreach (var adapter in adapters)
+            {
+                // Figure out if reader or writer
+                // Get component ID
+                // Store in data structures
+                Type requiredType = adapter.TypeOfMember;
+                var componentIdAttribute =
+                    (ComponentIdAttribute) Attribute.GetCustomAttribute(requiredType, typeof(ComponentIdAttribute), false);
+                if (componentIdAttribute == null)
+                {
+                    logger.HandleLog(LogType.Warning, new LogEvent(BadRequiredMemberWarning)
+                        .WithField(LoggingUtils.LoggerName, LoggerName)
+                        .WithField("MonoBehaviour", behaviourType.Name)
+                        .WithField("RequiredType", requiredType.Name));
+                    continue;
+                }
+
+                var componentId = componentIdAttribute.Id;
+                if (componentIdsToAdapters.ContainsKey(componentId))
+                {
+                    logger.HandleLog(LogType.Error, new LogEvent(MultipleReadersWritersRequiredError)
+                        .WithField(LoggingUtils.LoggerName, LoggerName)
+                        .WithField("MonoBehaviour", behaviourType.Name)
+                        .WithField("ComponentID", componentId)
+                        .WithField("RequiredType", requiredType.Name));
+                    badTypes.Add(behaviourType);
```

Comment:
Shouldn't this break this iteration of the loop? Currently we fire the error and then overwrite the previous adapter

Reply:
It doesn't really matter because whenever we would use the adapter afterwards, we first check and see that it's an invalid type and avoid touching it. But probably would be less confusing if we had a break, so I'll add that